File: ios/VineTrack/Backend/Repositories/OperationalToolPreferencesRepository.swift
import Foundation
import Supabase

/// The caller's saved Operational Tools layout (`sql/159`).
///
/// `hasPreference == false` means "no record yet" — the app must use the
/// VineTrack default order and show every authorised tool. A record is only
/// created once the user actually customises something.
nonisolated struct OperationalToolPreferences: Codable, Sendable, Equatable {
    let hasPreference: Bool
    let version: Int
    let visibleToolIds: [String]
    let hiddenToolIds: [String]
    let updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case hasPreference = "has_preference"
        case version
        case visibleToolIds = "visible_tool_ids"
        case hiddenToolIds = "hidden_tool_ids"
        case updatedAt = "updated_at"
    }

    init(hasPreference: Bool, version: Int, visibleToolIds: [String], hiddenToolIds: [String], updatedAt: Date?) {
        self.hasPreference = hasPreference
        self.version = version
        self.visibleToolIds = visibleToolIds
        self.hiddenToolIds = hiddenToolIds
        self.updatedAt = updatedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        hasPreference = (try? c.decodeIfPresent(Bool.self, forKey: .hasPreference)) ?? false
        version = (try? c.decodeIfPresent(Int.self, forKey: .version)) ?? 1
        visibleToolIds = (try? c.decodeIfPresent([String].self, forKey: .visibleToolIds)) ?? []
        hiddenToolIds = (try? c.decodeIfPresent([String].self, forKey: .hiddenToolIds)) ?? []
        // Timestamp formats vary between PostgREST configurations; the layout
        // must never fail to load because of a date string.
        updatedAt = try? c.decodeIfPresent(Date.self, forKey: .updatedAt)
    }

    static let empty = OperationalToolPreferences(
        hasPreference: false,
        version: 1,
        visibleToolIds: [],
        hiddenToolIds: [],
        updatedAt: nil
    )
}

/// Data source for the caller's Operational Tools layout. Abstracted so the
/// layout store can be unit-tested without a network round trip.
nonisolated protocol OperationalToolPreferencesProviding: Sendable {
    func fetch() async throws -> OperationalToolPreferences
    func save(visibleToolIds: [String], hiddenToolIds: [String]) async throws -> OperationalToolPreferences
    func reset() async throws -> OperationalToolPreferences
}

/// Reads and writes the caller's own Operational Tools layout. Every RPC is
/// SECURITY DEFINER and derives the user from `auth.uid()` — the layout is
/// per-user and shared across iOS, Android and the portal.
nonisolated struct OperationalToolPreferencesRepository: OperationalToolPreferencesProviding {
    private let provider: SupabaseClientProvider

    init(provider: SupabaseClientProvider = .shared) {
        self.provider = provider
    }

    func fetch() async throws -> OperationalToolPreferences {
        guard provider.isConfigured else {
            throw BackendRepositoryError.missingSupabaseConfiguration
        }
        return try await provider.client
            .rpc("get_my_operational_tool_preferences")
            .execute()
            .value
    }

    func save(visibleToolIds: [String], hiddenToolIds: [String]) async throws -> OperationalToolPreferences {
        guard provider.isConfigured else {
            throw BackendRepositoryError.missingSupabaseConfiguration
        }
        let params = SetPreferencesParams(
            visible: visibleToolIds,
            hidden: hiddenToolIds,
            version: 1
        )
        return try await provider.client
            .rpc("set_my_operational_tool_preferences", params: params)
            .execute()
            .value
    }

    func reset() async throws -> OperationalToolPreferences {
        guard provider.isConfigured else {
            throw BackendRepositoryError.missingSupabaseConfiguration
        }
        return try await provider.client
            .rpc("reset_my_operational_tool_preferences")
            .execute()
            .value
    }
}

nonisolated private struct SetPreferencesParams: Encodable, Sendable {
    let visible: [String]
    let hidden: [String]
    let version: Int

    enum CodingKeys: String, CodingKey {
        case visible = "p_visible_tool_ids"
        case hidden = "p_hidden_tool_ids"
        case version = "p_preference_version"
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/OperationalToolPreferencesRepository.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.auth.SessionStore
import io.ktor.client.call.body
import io.ktor.client.request.headers
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * The caller's saved Operational Tools layout as returned by SQL 159.
 *
 * `hasPreference == false` means "no record yet" — the app uses the VineTrack
 * default order and shows every authorised tool.
 */
@Serializable
data class OperationalToolPreferencesDto(
    @SerialName("has_preference") val hasPreference: Boolean = false,
    val version: Int = 1,
    @SerialName("visible_tool_ids") val visibleToolIds: List<String> = emptyList(),
    @SerialName("hidden_tool_ids") val hiddenToolIds: List<String> = emptyList(),
)

/**
 * Reads and writes the signed-in user's own Operational Tools layout.
 *
 * Every RPC is SECURITY DEFINER and derives the user from `auth.uid()`, so the
 * layout is per-user (never per-vineyard) and is shared with iOS and the
 * portal.
 */
class OperationalToolPreferencesRepository(private val session: SessionStore) {

    suspend fun fetch(): OperationalToolPreferencesDto = call("get_my_operational_tool_preferences", null)

    suspend fun save(
        visibleToolIds: List<String>,
        hiddenToolIds: List<String>,
    ): OperationalToolPreferencesDto = call(
        "set_my_operational_tool_preferences",
        buildJsonObject {
            put("p_visible_tool_ids", buildJsonArray { visibleToolIds.forEach { add(it) } })
            put("p_hidden_tool_ids", buildJsonArray { hiddenToolIds.forEach { add(it) } })
            put("p_preference_version", 1)
        },
    )

    suspend fun reset(): OperationalToolPreferencesDto = call("reset_my_operational_tool_preferences", null)

    private suspend fun call(rpc: String, body: JsonObject?): OperationalToolPreferencesDto =
        withContext(Dispatchers.IO) {
            if (!SupabaseClient.isConfigured) throw IllegalStateException("Supabase is not configured")
            val token = session.accessToken ?: throw IllegalStateException("Not signed in")
            val response = SupabaseClient.http.post(SupabaseClient.rpcUrl(rpc)) {
                headers {
                    append("apikey", SupabaseClient.anonKey)
                    append("Authorization", "Bearer $token")
                }
                contentType(ContentType.Application.Json)
                setBody(body ?: JsonObject(emptyMap()))
            }
            if (!response.status.isSuccess()) {
                // Sanitised: the server message never contains user data, but we
                // only surface the status to the UI layer.
                throw IllegalStateException("tool_preferences_rpc_failed_${response.status.value}")
            }
            val payload = response.body<JsonObject>()
            SupabaseClient.json.decodeFromJsonElement(
                OperationalToolPreferencesDto.serializer(),
                payload,
            )
        }
}
